osrf/vrx
# move tf to ros

This PR is an alternative to PR [664](https://github.com/osrf/vrx/pull/664) which also resolves many of the TF issues in [681](https://github.com/osrf/vrx/issues/681).  It changes the way we publish TF information on the ROS side by:
* adding a JointStatePublisher plugin to the wamv and bridging the topic to ROS to make dynamic joint states available on the ROS side
* modifying the PosePublisher plugin to only publish sensor transforms, as those are only available via the sdf and will not be published when parsing the urdf
* replacing most of the functionality in the custom pose_tf_broadcaster node with a robot_state_publisher node, which provides the robot description and tf tree in a format that can be ingested by ROS applications such as RViz

A launch file to start an instance of RViz with a custom config file is included for testing and for the tutorial.

UNKNOWN, help needed: Is any other process on the Gazebo side using the transforms from the sdf that PosePublisher was producing previously?  I think not, as we have access to the truth data, but if so this PR will likely break something and needs to be reworked.

To test:

`ros2 launch vrx_gz competition.launch.py world:=stationkeeping_task`
`ros2 launch vrx_gazebo rviz.launch.py`

Confirm the wamv appears with all components drawn and in the right place.

`ros2 topic pub /wamv/thrusters/left/pos std_msgs/msg/Float64 "{data: -1}"`

Watch the left engine/propeller move in RViz as it moves in Gazebo.

`ros2 run tf2_tools view_frames`

Confirm the new TF tree is structured correctly.

I still haven't looked into the extra "wamv/" prepended on the PosePublisher's output, so this code adds the prefix to all ROS-generated link names instead, in order to make the sensor sdf links connect correctly.  This is minor but annoying, so if we decide to implement this I will take another look at it.
